### PR TITLE
Sample test coverage for SSR + hydration

### PR DESCRIPTION
**Pull request by Shinoni**

### CODE REVIEW

File: example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js [hunk 1]
```diff
@@ -1,47 +1,55 @@
...
+        expect(componentEl?.shadowRoot).not.toBeNull();
+
+        // Hydrate the component with the provided props
         hydrateComponent(componentEl, Basic, props);
 
-        // Query the h1 element inside the shadow root
-        const shadowRoot = componentEl.shadowRoot;
-        expect(shadowRoot).not.toBeNull();
+        // Validate that the shadow DOM content matches the pre-hydration markup
+        expect(componentEl.shadowRoot.innerHTML).toBe(preHydratedMarkup);
```
Comment (lpomerleau) [inline example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js:47]: This is a great thing to test. Can you make a comment that it is a simple [Cumulative Layout Shifts (CLS)](https://web.dev/articles/cls) test? CLS (along with hydration errors) is important to check for performance.

File: example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js [hunk 2]
```diff
@@ -1,47 +1,55 @@
...
         hydrateShadowRoots(wrapper);
 
         const componentEl = wrapper.firstElementChild;
-        expect(componentEl).toBeInstanceOf(HTMLElement);
-        expect(componentEl).toHaveProperty('shadowRoot');
 
+        // Check that the component is an instance of HTMLElement and has a shadow root
+        expect(componentEl?.shadowRoot).not.toBeNull();
+
+        // Hydrate the component with the provided props
```
Comment (lpomerleau) [inline example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js:43]: How do we check for hydration errors? Does [jest.ssr-client.setupAfterEnv.js](https://github.com/salesforce/lwc-test/blob/e43ba7be692b42ce91c992828547a3749df1a932/example-ssr/jest.ssr-client.setupAfterEnv.js) do this automatically for all `ssr-client` test suites?

File: example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js [hunk 3]
```diff
@@ -1,47 +1,55 @@
-import { readSnapshotMarkup } from '@lwc/jest-ssr-snapshot-utils';
-import { hydrateShadowRoots } from '@webcomponents/template-shadowroot';
-import { hydrateComponent } from '@lwc/engine-dom';
-import Basic from '../basic';
-import tests from './ssr';
+const { readSnapshotMarkup } = require('@lwc/jest-ssr-snapshot-utils');
+const { hydrateShadowRoots } = require('@webcomponents/template-shadowroot');
+const { hydrateComponent } = require('@lwc/engine-dom');
+const Basic = require('../basic').default;
+const tests = require('./test-data').default;
```
[the remark this answers is not in the document]
Comment (Shinoni) [reply example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js:5]: Just for consistency with other tests in this repo.

File: example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js [hunk 4]
```diff
@@ -1,47 +1,55 @@
...
+        // Set the inner HTML of the wrapper to the snapshot markup
         wrapper.innerHTML = markup;
 
-        // Hydrate shadow roots and component
+        // Extract the content inside the template element (pre-hydration)
+        const preHydratedMarkup = wrapper.querySelector(
+            'template[shadowrootmode="open"]'
+        ).innerHTML;
+
+        // Hydrate the shadow roots for the wrapper
```
[the remark this answers is not in the document]
Comment (Shinoni) [reply example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js:35]: Sure.

File: example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js (hunk 2 above)
Comment (Shinoni) [reply example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js:43]: Yes, that's the only way that was suggested by Nolan.

File: example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (lpomerleau) [reply example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js:5]: I'd recommend sticking with ESM for a few reasons:
- it's the modern syntax
- the LWC components themselves use ESM
- falling back to CommonJS is not needed for these tests to be functional

File: example-ssr/README.md [hunk 5]
```diff
@@ -1,114 +1,205 @@
-# SSR component Unit test setup
+# SSR Testing Strategy
```
Comment (lpomerleau) [inline example-ssr/README.md:1]: Is this file a copy of [the quip doc](https://salesforce.quip.com/70uGAbzZ4Tg3)? I'd recommend picking 1 and linking to the other, so you don't have to keep both in sync.

Comment (Shinoni) [reply example-ssr/README.md:1]: Yes, for now. The Quip document is intended for collaboration with other stakeholders, and later I'll link it to the README file.

File: example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Shinoni) [reply example-ssr/src/modules/x/basic/__tests__/basic-data-driven.ssr-client.test.js:5]: Sounds good!

File: example-ssr/README.md (hunk 5 above)
Comment (lpomerleau) [reply example-ssr/README.md:1]: Sounds good! I like when the doc is close to the code

**Review by lpomerleau — APPROVED** [review]

**Pull request merged**